### PR TITLE
Fix wrong result query during MergeJoin

### PR DESCRIPTION
**Pull request by hardikar**
Consider the query and corresponding plan below : 

```
shardikar=# explain select * FROM gp_dist_random('alpha') FULL OUTER JOIN gp_dist_random('theta') ON (alpha.i = theta.i) WHERE (alpha.j IS NULL or theta.j ISNULL);
                                                   QUERY PLAN
-----------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)  (cost=1000377240.02..1000377240.14 rows=5 width=40000016)
   ->  Merge Full Join  (cost=1000377240.02..1000377240.14 rows=2 width=40000016)
         Merge Cond: alpha.i = theta.i
         Filter: alpha.j IS NULL OR theta.j IS NULL
         ->  Sort  (cost=1.26..1.27 rows=2 width=8)
               Sort Key: alpha.i
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..1.18 rows=2 width=8)
                     Hash Key: alpha.i
                     ->  Seq Scan on alpha  (cost=0.00..1.06 rows=2 width=8)
         ->  Materialize  (cost=377238.76..377238.83 rows=2 width=40000008)
               ->  Sort  (cost=377238.76..377238.77 rows=2 width=40000008)
                     Sort Key: theta.i
                     ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..1.18 rows=2 width=40000008)
                           Hash Key: theta.i
                           ->  Seq Scan on theta  (cost=0.00..1.06 rows=2 width=40000008)
 Optimizer status: legacy query optimizer
(16 rows)
```

When the size of the inner tuple in a MergeJoin is large, planner
decides to place a Materialize on the top of the inner Sort operator.
Furthemore if there is any Motion below the MergeJoin,
mjnode->prefetch_inner is set to true.

During execution of plan that has a Motion underneath a Materialize
underneath a MJ, MJ retrieves one tuple from the Motion (via
Materialize) and then resets the stream (by calling ExecReScan). Since
the Materialize underneath is initialized with incorrect eflags, it
deletes its tuplestore, instead of rewinding its cursor.  This fix sets
the right eflags value during Init (see comments in ExecMaterialReScan).

@foyzur @hlinnaka @karthijrk Please take a look.

### CODE REVIEW

File: src/test/regress/sql/join.sql [hunk 1]
```diff
@@ -528,3 +528,25 @@ select q1, unique2, thousand, hundred
...
+CREATE TABLE theta (i int, j char(10000000));
+
+INSERT INTO alpha values (1, 1), (2, 2);
+INSERT INTO theta values (1, 'f'), (2, 'g');
+
+SELECT *
+FROM gp_dist_random('alpha') FULL OUTER JOIN gp_dist_random('theta')
+  ON (alpha.i = theta.i)
+WHERE (alpha.j IS NULL or theta.j IS NULL);
+
```
Comment (vraghavan78) [inline src/test/regress/sql/join.sql]: * I think this will conflict with Postgres merge. Please add the test in join_gp.sql
* How do you plan to ensure for this test that planner always picks MergeJoin? You may want to consider the GUC to enable merge join and disable other joins (hash and nlj). If you do use those GUCs then please reset them after the test.

File: src/backend/executor/nodeMergejoin.c [hunk 2]
```diff
@@ -1585,6 +1586,7 @@ ExecInitMergeJoin(MergeJoin *node, EState *estate, int eflags)
 
 	mergestate->prefetch_inner = node->join.prefetch_inner;
 	mergestate->mj_squelchInner = true;
+	rewindflag = mergestate->prefetch_inner ? EXEC_FLAG_REWIND : 0;
```
Comment (vraghavan78) [inline src/backend/executor/nodeMergejoin.c]: Can you add a comment here please to explain why this is done?

File: src/test/regress/sql/join.sql (hunk 1 above)
Comment (hardikar) [inline src/test/regress/sql/join.sql]: * Sure.
* According to the code, planner always picks MJ for a FULL OUTER JOIN (even setting enable_mergejoin=off). But I will add those in any case.

**Comment by hlinnaka** [comment]
I don't understand what prefetch_inner does. The comments and MPP-3300 suggests that there's some kind of a deadlock hazard if both sides of a join contain a Motion node, but I don't see any further explanation of how the deadlock might occur.

Also, isn't it pointless to have a Materialize node on top of a Sort node? A Sort materializes the output anyway (or at least can do so, if it's marked as randomAccess).

**Comment by hlinnaka** [comment]
> Also, isn't it pointless to have a Materialize node on top of a Sort node? A Sort materializes
> the output anyway (or at least can do so, if it's marked as randomAccess).

@hardikar explained this to me off-list:

There's this comment in createplan.c:
```
	/*
	 * If inner plan is a sort that is expected to spill to disk, add a
	 * materialize node to shield it from the need to handle mark/restore.
	 * This will allow it to perform the last merge pass on-the-fly, while in
	 * most cases not requiring the materialize to spill to disk.
	 *
	 * XXX really, Sort oughta do this for itself, probably, to avoid the
	 * overhead of a separate plan node.
	 */
	if (IsA(inner_plan, Sort) &&
		sort_exceeds_work_mem((Sort *) inner_plan))
	{
		Plan	   *matplan = (Plan *) make_material(inner_plan);
```

So adding the Materialize allows the merge step of the sort to happen "on-the-fly". That doesn't affect the total cost, but it allows returning tuples earlier, before completing the sort in full, which gives a better start up cost. Makes sense.

**Comment by hardikar** [comment]
> I don't understand what prefetch_inner does. The comments and MPP-3300 suggests that there's some kind of a deadlock hazard if both sides of a join contain a Motion node, but I don't see any further explanation of how the deadlock might occur.

According to the code, `prefetch_inner` tries to retrieve at least one tuple from the inner node, then calls ReScan(innerPlan) resetting the input stream before starting with the outer loop (in case of MJ). After digging a while into our archived history, I couldn't find a cohesive explanation as to why `prefetch_inner` prevents a deadlock.

MPP-3300 refers to the "deadlock hazard" elog messages from [nodeMotion.c](https://github.com/greenplum-db/gpdb/blob/ef42590d2e5f9d01ee5839d93b1d6b36fd2b1e9e/src/backend/executor/nodeMotion.c#L23). Related to the fix is `motion_sanity_check()` method which traverses the tree making sure that the plan is safe from MOTION DEADLOCKs. This mainly concerned about joins with Motion nodes in the left and right subtrees.
```
			/* if this node is "deadlock safe" then even if we have
			 * motion on both sides we will not deadlock (because the
			 * access pattern is deadlock safe: all rows are retrieved
			 * from one side before the first row from the other). */
			if (!deadlock_safe && ((left_result.flags & SANITY_MOTION) &&
								   (right_result.flags & SANITY_MOTION)))
			{
				elog(LOG, "FOUND MOTION DEADLOCK in %s", branch_label);
				result->flags |= SANITY_DEADLOCK;
				break;
			}
```

In `create_mergejoin_plan()`, there is a check for motion hazards on both subtrees, and if so it sets prefetch_inner to true and add a slackening operator (Materialize) on top of the inner Motion. (see in createplan.c). Searching for `node->motionHazard`, I see that it is set to true when there is a Motion, CTEScans, SubQueryScans, and TableFunctionScans node in the subplan.


In summary, there seems to be a possibility of a deadlock when there is a Motion in both subtrees of a Join, but I do not know enough about Motion to say how with certainty. The fix seems to be to add a prefetch the inner plan. There is something similar for NLJ also. 
I should also add that I *did* find references to a deadlock that can happen when there is a SharedInputScan with two producers and two consumers in different slices. (see [OPT-2690](https://github.com/greenplum-db/gpdb/blob/ef42590d2e5f9d01ee5839d93b1d6b36fd2b1e9e/src/backend/executor/nodeShareInputScan.c#L681)). But this is fixed and may or may not be related to prefetch_inner.

**Comment by hardikar** [comment]
In any case, from the perspective of query execution, when `prefetch_inner = true`, MJ calls Rescan(Materialize) and materialize expects EXEC_FLAG_REWIND to support Rescan:
	/*
		 * If subnode is to be rescanned then we forget previous stored
		 * results; we have to re-read the subplan and re-store.  Also, if we
		 * told tuplestore it needn't support rescan, we lose and must
		 * re-read.  (This last should not happen in common cases; else our
		 * caller lied by not passing EXEC_FLAG_REWIND to us.)
		 *
		 * Otherwise we can just rewind and rescan the stored output. The
		 * state of the subnode does not change.
		 */
		if (((PlanState *) node)->lefttree->chgParam != NULL ||
			(node->eflags & EXEC_FLAG_REWIND) == 0)
		{
```

Thoughts?

**Comment by Eulerizeit** [comment]
@ashwinstar this was the PR I was referring to this morning.

**Comment by vraghavan78** [comment]
@hlinnaka thanks for the CTE scan example on the https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/gMa1tW0x_fk

So can @hardikar merge this PR and QO folks look into the CTE issue and another planner related issue?

**Comment by hlinnaka** [comment]
I think we should get to the bottom of this first.

**Comment by hardikar** [comment]
Added comments about how the deadlock happens in https://github.com/greenplum-db/gpdb/pull/1642. 
So since we need the prefetch_inner in such cases, we need an execReScan which means we need to send the right eflags to Materialize underneath the MJ. Does this PR look good to you then @hlinnaka ?

**Comment by hlinnaka** [comment]
Ok, looks good to me then.
